control-plane/connect-inject/endpoints_controller.go

```diff
@@ -834,6 +835,168 @@ func serviceInstancesForK8SServiceNameAndNamespace(k8sServiceName, k8sServiceNam
+	var preparedQuery string
+	var port int32
+	parts := strings.SplitN(rawUpstream, ":", 3)
+
+	port, _ = portValue(pod, strings.TrimSpace(parts[2]))
+	preparedQuery = strings.TrimSpace(parts[1])
+	var upstream api.Upstream
+	if port > 0 {
+
+		upstream = api.Upstream{
+			DestinationType: api.UpstreamDestTypePreparedQuery,
+			DestinationName: preparedQuery,
+			LocalBindPort:   int(port),
+		}
+	}
+	return upstream
+}
+
+func (r *EndpointsController) processNonAnnotatedUpstream(pod corev1.Pod, rawUpstream string) (api.Upstream, error) {
+	var datacenter, serviceName, namespace, partition, peer string
+	var port int32
+	var upstream api.Upstream
+
+	parts := strings.SplitN(rawUpstream, ":", 3)
+
+	port, _ = portValue(pod, strings.TrimSpace(parts[1]))
+
+	// If Consul Namespaces or Admin Partitions are enabled, attempt to parse the
+	// upstream for a namespace.
+	if r.EnableConsulNamespaces || r.EnableConsulPartitions {
+		pieces := strings.SplitN(parts[0], ".", 3)
+		switch len(pieces) {
+		case 3:
+			partition = strings.TrimSpace(pieces[2])
+			fallthrough
+		case 2:
+			namespace = strings.TrimSpace(pieces[1])
+			fallthrough
+		default:
+			serviceName = strings.TrimSpace(pieces[0])
+		}
+	} else {
+		serviceName = strings.TrimSpace(parts[0])
+	}
+
+	// parse the optional datacenter
+	if len(parts) > 2 {
+		datacenter = strings.TrimSpace(parts[2])
+
+		// Check if there's a proxy defaults config with mesh gateway
+		// mode set to local or remote. This helps users from
+		// accidentally forgetting to set a mesh gateway mode
+		// and then being confused as to why their traffic isn't
+		// routing.
+		entry, _, err := r.ConsulClient.ConfigEntries().Get(api.ProxyDefaults, api.ProxyConfigGlobal, nil)
+		if err != nil && strings.Contains(err.Error(), "Unexpected response code: 404") {
+			return api.Upstream{}, fmt.Errorf("upstream %q is invalid: there is no ProxyDefaults config to set mesh gateway mode", rawUpstream)
+		} else if err == nil {
+			mode := entry.(*api.ProxyConfigEntry).MeshGateway.Mode
+			if mode != api.MeshGatewayModeLocal && mode != api.MeshGatewayModeRemote {
+				return api.Upstream{}, fmt.Errorf("upstream %q is invalid: ProxyDefaults mesh gateway mode is neither %q nor %q", rawUpstream, api.MeshGatewayModeLocal, api.MeshGatewayModeRemote)
+			}
+		}
+		// NOTE: If we can't reach Consul we don't error out because
+		// that would fail the pod scheduling and this is a nice-to-have
+		// check, not something that should block during a Consul hiccup.
+	}
+	if port > 0 {
+		upstream = api.Upstream{
+			DestinationType:      api.UpstreamDestTypeService,
+			DestinationPartition: partition,
+			DestinationPeer:      peer,
+			DestinationNamespace: namespace,
+			DestinationName:      serviceName,
+			Datacenter:           datacenter,
+			LocalBindPort:        int(port),
+		}
+	}
+	return upstream, nil
+
+}
+
+func (r *EndpointsController) processAnnotatedUpstream(pod corev1.Pod, rawUpstream string) (api.Upstream, error) {
```

Comment:
Can we copy the comment where you have described the formatting of the annotated upstream and add that above each of these method to describe the "shape" of the upstream that they will each process

Reply:
Yup added!